RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc

```diff
@@ -970,3 +653,240 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
     theEvent.put(std::move(gsfElectronPFClusterIsos[iid]),outGsfElectronPFClusterIsos_[iid]);
   }  
 }
+
+void ReducedEGProducer::linkPhotonCore(const reco::Photon& photon, 
+				       reco::PhotonCoreCollection& photonCores, 
+				       std::map<reco::PhotonCoreRef, unsigned int>& photonCoreMap)
+{
+  const reco::PhotonCoreRef &photonCore = photon.photonCore();
+  if (!photonCoreMap.count(photonCore)) {
+    photonCores.push_back(*photonCore);
+    photonCoreMap[photonCore] = photonCores.size() - 1;
+  }
+}
+    
+void ReducedEGProducer::linkSuperCluster(const reco::SuperClusterRef& superCluster, 
+					 std::map<reco::SuperClusterRef, unsigned int>& superClusterMap,
+					 reco::SuperClusterCollection& superClusters,
+					 const bool relink,
+					 std::unordered_set<unsigned int>& superClusterFullRelinkMap)
+{
+  const auto &mappedsc = superClusterMap.find(superCluster);
+  //get index in output collection in order to keep track whether superCluster
+  //will be subject to full relinking
+  unsigned int mappedscidx = 0;
+  if (mappedsc==superClusterMap.end()) {
+    superClusters.push_back(*superCluster);
+    mappedscidx = superClusters.size() - 1;
+    superClusterMap[superCluster] = mappedscidx;
+  }
+  else {
+    mappedscidx = mappedsc->second;
+  }
+
+  //additionally mark supercluster for full relinking
+  if (relink) superClusterFullRelinkMap.insert(mappedscidx);
+}
+
+
+void ReducedEGProducer::linkConversions(const reco::ConversionRefVector& convrefs,
+					reco::ConversionCollection& conversions, 
+					std::map<reco::ConversionRef, unsigned int>& conversionMap)
+{
+  for (const reco::ConversionRef &convref : convrefs) {
+    linkConversion(convref, conversions, conversionMap);
+  }
+}
+
+void ReducedEGProducer::linkConversionsByTrackRef(const edm::Handle<reco::ConversionCollection>& conversionHandle,
+						  const reco::GsfElectron& gsfElectron,
```

Comment:
actually, this is one we cannot make a template of, as the matchesConversion() implementations are different based on the derived type (L710 vs 726).  That said, there are quite a few other places for templates, which I implementing.